Fix a potential compiler bug for CUDA-aware MPI call

sjsprecious opened the pull request:
This PR fixes a potential compiler bug while invoking the `mpi_allreduce` for communication.

**Issue:**
- It turns out that the `!$acc update host` and `!$acc update device` directives around an `MPI_ALLREDUCE` call do not work properly and produce a wrong result in a GPU run unless the managed memory is used.

**Solution:**
- use the "!$acc host_data use_device" directive
- specify the array bounds explicitly in the "!$acc update host" and "!$acc update device" directives

Both solutions work so far and we decide to go with the first option since it enables the CUDA-aware MPI communication. No managed memory is needed for the new GPU run.

**Verification result:**

GPU run with managed memory (from https://github.com/george-bryan/CM1/pull/88):
- Identical variables: 44
- variables with red_diff > 1.e-6: 29
- variables with red_diff <= 1.e-6: 13
- four variables with largest error: KSVMAX, KSHMAX, VOR2KM, VOR1KM

GPU run without managed memory:
- Identical variables: 44
- variables with red_diff > 1.e-6: 29
- variables with red_diff <= 1.e-6: 13
- four variables with largest error: KSVMAX, KSHMAX, VOR2KM, VOR1KM

sjsprecious committed:
Turn off managed memory
Replace "acc update host/device" with "acc host_data use_device" for array variables

	modified:   Makefile
	modified:   domaindiag.F
	modified:   testcase_simple_phys.F
	modified:   turb.F
	modified:   writeout.F

johnmauff commented on src/Makefile:
```diff
@@ -116,7 +116,7 @@ ifeq (${FC},nvfortran)
             else
                 OPTS += -acc -gpu=cc70,lineinfo,nofma,math_uniform -Mpcast -Minfo=accel
-                OPTS_M += -acc -gpu=cc70,lineinfo,nofma,math_uniform,managed -Mpcast -Minfo=accel
+                OPTS_M += -acc -gpu=cc70,lineinfo,nofma,math_uniform -Mpcast -Minfo=accel
```
Do we want to maintain an OPTS_M compiler flag in the Makefile now that we do not need the use of Managed memory?

sjsprecious replied:
Thanks @johnmauff. This is a good point. I keep it here in case we want to apply separate flags to different files in the future. But it does seem unnecessary by now since we no longer need the managed memory option. Shall we just remove it?

sjsprecious replied:
The Makefile has been updated to remove the `OPTS_M` option.

johnmauff reviewed (commented):
Thanks for submitting this PR.  It looks good with the potential exception of the issue in the Makefile.

johnmauff commented:
@sjsprecious, I noticed that there are other '!$acc update host()' that are associated with MPI reduction calls.  For example in solve1.F and solve2.F. Is there a particular reason why these were not modified?

sjsprecious commented:
> @sjsprecious, I noticed that there are other '!$acc update host()' that are associated with MPI reduction calls. For example in solve1.F and solve2.F. Is there a particular reason why these were not modified?

Thanks for catching them. I do not modify them because it is a `mpi_iallreduce` call and thus skips from my `grep` search. Let me go through the code again and eliminate the potentially buggy `!$acc update host` directive.

sjsprecious committed:
Address John's comments for PR

	modified:   Makefile
	modified:   misclibs.F
	modified:   solve1.F
	modified:   solve2.F
	modified:   writeout.F

sjsprecious commented:
@johnmauff I have removed additional `!$acc update host()` directives in other source code. The verification results remain unchanged. Let me know if you have further questions.

johnmauff commented:
@sjsprecious Thanks for the updates.  It now looks good.

sherimickelson reviewed (approved):
Thanks @sjsprecious for adding in these changes
